### PR TITLE
feat(relay): don't close connections upon errors in relay server

### LINKED ISSUE #4716: Remove usage of `ConnectionHandlerEvent::Close` from `libp2p_relay`
As part of https://github.com/libp2p/rust-libp2p/issues/3591 we need to stop closing connections upon errors on individual streams.

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

To remove the usages of `ConnectionHandlerEvent::Close` from the relay-server, we unify what used to be called `CircuitFailedReason` and `FatalUpgradeError`. Whilst the errors may be fatal for the particular circuit, they are not necessarily fatal for the entire connection.

Related: #3591.
Resolves: #4716.

## Notes & open questions

- ~Should we do some kind of "smart" connection management upon failures on the streams further up? At the moment, we don't expose the details of which connection a stream failed on. I am leaning towards saying "no" here and instead relying more on https://github.com/libp2p/rust-libp2p/pull/4595. Once we do more automated keep-alive tracking, bad connections will close automatically much more aggressively. That is because any error on a stream will lead to the user dropping the stream which means we will automatically return `KeepAlive::No`.~

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/relay/CHANGELOG.md
Comment (thomaseizinger) [inline protocols/relay/CHANGELOG.md]:
```suggestion
  See [PR 4718](https://github.com/libp2p/rust-libp2p/pull/4718).
```

File: protocols/relay/src/behaviour.rs
Comment (thomaseizinger) [inline protocols/relay/src/behaviour.rs:178]: @mxinden Should we include the `ConnectionId` here so that users _could_ close the connection if they wanted to?

Comment (mxinden) [reply protocols/relay/src/behaviour.rs:178]: :+1: sounds good. For the sake of consistency, how about including it in every event variant?

File: protocols/relay/src/behaviour/handler.rs
Comment (mxinden) [inline protocols/relay/src/behaviour/handler.rs:715]: I would say so. What alternatives do you have in mind?

**Review by mxinden — COMMENTED** [review]
Thank you for the work.

Can you add a test that ensures https://github.com/libp2p/rust-libp2p/discussions/4752 is fixed with this pull request, i.e. that a connection is not closed even though the remote does not support the relay protocol as one expected?

File: protocols/relay/src/behaviour.rs
Comment (thomaseizinger) [reply protocols/relay/src/behaviour.rs:178]: Is it worth the complexity though? The more I am working through the codebase and thinking about the error cases, the less I am convinced that we should emit errors for something like that at all.

In this particular case, all of these errors are essentially protocol violations by the remote peer. I wouldn't know what code to write in response to such an error. If all streams fail, the keep alive algorithm will close the connection anyway.

The only case I can think of that maybe justifies keep these events are the metrics. But the value of those is also questionable. Why do I need _metrics_ on how many "broken" requests I am getting? This is what logs are for, no? A node operator can (should?) create alerts on certain logs. These errors are certainly unexpected so perhaps we should just log them as `warn!` and be done with it?

**Comment by thomaseizinger** [comment]
> Can you add a test that ensures #4752 is fixed with this pull request, i.e. that a connection is not closed even though the remote does not support the relay protocol as one expected?

To properly test this, I think we need to also merge the fix for the client side (https://github.com/libp2p/rust-libp2p/pull/4745) otherwise, the client will simply close the connection. Happy to add a test once both PRs are landed.

File: protocols/relay/src/behaviour.rs
Comment (thomaseizinger) [reply protocols/relay/src/behaviour.rs:178]: I opted for deprecating them for now and open an issue where users can weigh in if they need the events for something else. If nobody speaks up until the next breaking release, I'd suggest to move forward with removing them.

See https://github.com/libp2p/rust-libp2p/issues/4757.

File: protocols/relay/src/behaviour/handler.rs
Comment (thomaseizinger) [reply protocols/relay/src/behaviour/handler.rs:715]: No alternatives.

File: protocols/relay/src/behaviour.rs
Comment (mxinden) [reply protocols/relay/src/behaviour.rs:178]: > Why do I need _metrics_ on how many "broken" requests I am getting? This is what logs are for, no? A node operator can (should?) create alerts on certain logs. These errors are certainly unexpected so perhaps we should just log them as `warn!` and be done with it?

I would expect users to set up alerting via metrics, not logs.

I do believe metrics on errors are important. E.g. when error rates suddenly spike, that is worth alerting on.

Agreed that exposing the errors just for `libp2p-metrics` is unfortunate. Long term I would like to integrate `prometheus-client` into `libp2p-*` directly.

Comment (thomaseizinger) [reply protocols/relay/src/behaviour.rs:178]: > > Why do I need _metrics_ on how many "broken" requests I am getting? This is what logs are for, no? A node operator can (should?) create alerts on certain logs. These errors are certainly unexpected so perhaps we should just log them as `warn!` and be done with it?
> 
> I would expect users to set up alerting via metrics, not logs.
> 
> I do believe metrics on errors are important. E.g. when error rates suddenly spike, that is worth alerting on.

If we'd follow that logic, we shouldn't be logging any errors or warnings, right?

I don't think that solution scales. I am no SRE but it sounds painful to configure alerts based on many different metrics when the alternative is as easy as: "Alert me if there is a sudden increase of `warn` logs"?

As I've argued in the issue, I think metric based alerts are super powerful if you are tracking those metrics anyway. But just creating them so you can set an alert when they show up seems like a lot of unnecessary effort, both for us as maintainers and for users.

Can we merge it with the warnings or are you fundamentally opposed to it?

**Review by mxinden — APPROVED** [review]

**Pull request merged**
